Execute Validation Plans in Device Ingest

lseppala opened the pull request:
This work adds the following:

* Adds validation results to the schema and association with validation state.
* Creates 2 validation plans (one each for servers and switches) corresponding to the legacy validation logic.
* Removes existing Minion code. We decided that it's not necessary to run it asynchronously from a queue at this time, as it adds operational complexity.
* Adds `Conch::Model::ValidationState->run_validation_plan`, which creates a new validation state, executes a validation plan for a device, and stores new _unique results not in the latest validation state and reuses the non-unique results.
* Executes `run_validation_plan` in the device report ingest pipeline. This is guarded with a feature-flag. The objective is to run both the new and old validations in parallel to evaluate effectiveness and build history.

sungo commented on Conch/cpanfile:
```diff
@@ -46,6 +46,7 @@ requires 'String::CamelCase', '>= 0.04';
 
 # database
 requires 'DBD::Pg';
+requires 'DateTime::Format::Pg';
```
I thought we decided to not use this module any more. I'm not seeing where it's used in the PR either.

lseppala replied:
I got an exception for this module missing during evaluation of the old validation logic. It's still used as part of the DBIC classes in Conch::Legacy::Schema::* through DBIx::Class::InflateColumn::DateTime.

We may not have seen this exception in production because DateTime::Format::Pg is still in the `local` cache from previous deployments.  I had recently blown away my `local` dir when it occurred

I'll reproduce the issue and add comments.

lseppala replied:
I removed this dependency, deleted my `local` dir, `carton install`, and processed a report. It threw an exception missing the dependency.

```
[Thu Apr  5 09:22:59 2018] [error] Can't locate DateTime/Format/Pg.pm in @INC (you may need to install the DateTime::Format::Pg module) (@INC contains: /Users/lsep/dev/joyent/conch/Conch/bin/../lib /Users/lsep/dev/joyent/conch/Conch/local/lib/perl5/darwin-thread-multi-2level /Users/lsep/dev/joyent/conch/Conch/local/lib/perl5 /usr/local/Cellar/perl/5.26.1/lib/per
l5/site_perl/5.26.1/darwin-thread-multi-2level /usr/local/Cellar/perl/5.26.1/lib/perl5/site_perl/5.26.1 /usr/local/Cellar/perl/5.26.1/lib/perl5/5.26.1/darwin-thread-multi-2level /usr/local/Cellar/perl/5.26.1/lib/perl5/5.26.1 /usr/local/lib/perl5/site_perl/5.26.1) at /Users/lsep/dev/joyent/conch/Conch/local/lib/perl5/Class/C3/Component
ised.pm line 146.
141:     return if ( *{"${f_class}::$_"}{CODE} );
142:   }
143:
144:   # require always returns true on success
145:   # ill-behaved modules might very well obliterate $_
146:   eval { local $_; require($file) } or do {
147:
148:     $@ = "Invalid class name '$f_class'" if $f_class =~ $invalid_class;
149:
150:     if ($class->can('throw_exception')) {
151:       $class->throw_exception($@);

[Thu Apr  5 09:22:59 2018] [error] Stack Trace (first 10 frames):
        Mojolicious::__ANON__ at /Users/lsep/dev/joyent/conch/Conch/local/lib/perl5/Class/C3/Componentised.pm:146
        (eval) at /Users/lsep/dev/joyent/conch/Conch/local/lib/perl5/Class/C3/Componentised.pm:146
        Class::C3::Componentised::ensure_class_loaded at /Users/lsep/dev/joyent/conch/Conch/local/lib/perl5/DBIx/Class/AccessorGroup.pm:19
        DBIx::Class::AccessorGroup::get_component_class at (eval 1156):2
        DBIx::Class::Storage::DBI::datetime_parser_type at /Users/lsep/dev/joyent/conch/Conch/local/lib/perl5/DBIx/Class/Storage/DBI.pm:3141
        DBIx::Class::Storage::DBI::build_datetime_parser at /Users/lsep/dev/joyent/conch/Conch/local/lib/perl5/DBIx/Class/Storage/DBI.pm:3124
        DBIx::Class::Storage::DBI::datetime_parser at /Users/lsep/dev/joyent/conch/Conch/local/lib/perl5/DBIx/Class/InflateColumn/DateTime.pm:221
        DBIx::Class::InflateColumn::DateTime::_datetime_parser at /Users/lsep/dev/joyent/conch/Conch/local/lib/perl5/DBIx/Class/InflateColumn/DateTime.pm:196
        DBIx::Class::InflateColumn::DateTime::_flate_or_fallback at /Users/lsep/dev/joyent/conch/Conch/local/lib/perl5/DBIx/Class/InflateColumn/DateTime.pm:212
        DBIx::Class::InflateColumn::DateTime::_inflate_to_datetime at /Users/lsep/dev/joyent/conch/Conch/local/lib/perl5/DBIx/Class/InflateColumn/DateTime.pm:175
        DBIx::Class::InflateColumn::DateTime::__ANON__ at /Users/lsep/dev/joyent/conch/Conch/local/lib/perl5/DBIx/Class/InflateColumn.pm:122
```

Oddly, DateTime formats for SQLite and MySQL are pre-requistes for  DBIx::Class::TimeStamp and installed, but not for Postgres.

```
...
Successfully installed DateTime-Format-SQLite-0.11
Successfully installed DateTime-Format-MySQL-0.06
...
```

sungo commented on Conch/lib/Conch/Model/Validation.pm:
```diff
@@ -17,10 +17,28 @@ use Conch::Model::DeviceLocation;
+		module      => 'Conch::Validation::ExampleValidation',
+	);
+
+All unspecied fields will be 'undef'.
```
typo

sungo commented on Conch/lib/Conch/Validation.pm:
```diff
@@ -622,6 +644,23 @@ sub clear_results ( $self ) {
 
+# Internal error representation. not used outside of this module. Extends
+# 'Mojo::Exception' to store a 'hint' attribute.
+package Conch::Validation::Error;
```
Please break this out into a separate file

lseppala replied:
Honest question. I intended fully for this exception to only be used within Conch::Validation and this was the best approximation of a 'private class' as you'd find in other languages. Is there a particular reason to do this in Perl or is it a preference?

sungo replied:
Modules in their own file is convention, yes, but it's a sensible one because it makes finding code easier. When working with the code, I will expect to find Conch::Validation::Error in lib/Conch/Validation/Error.pm and will waste time hunting for it otherwise.

sungo commented on Conch/lib/Conch/ValidationSystem.pm:
```diff
@@ -71,95 +70,76 @@ sub load_validations ( $class, $logger = Mojo::Log->new ) {
-			my @results = map { $_->output_hash }
-				$validation->run_validation_for_device( $device, $data )->@*;
-			$job->finish( \@results );
+	my @server_validations = qw( bios_firmware_version cpu_count cpu_temperature
```
Should this maybe be out in a config someplace?

lseppala replied:
I'll do this. I initially decided against it because I hope to drop these legacy validation plans soon as we move to the new orchestration/validation world, but it's entirely likely that we might add new validations to these plans before we fully migrate over.

sungo commented on sql/migrations/0025-validation-results.sql:
```diff
@@ -0,0 +1,27 @@
+select run_migration(25, $$
```
the PR I submitted yesterday uses migration number 25

sungo commented:
Tests are failing

```
t/validations/00_base_class.t .......................... 1/? 
    #   Failed test 'threw Conch::Validation::Error'
    #   at t/validations/00_base_class.t line 44.
    # expecting: Conch::Validation::Error
    # found: Conch::ValidationError (Validation dies)
    # Looks like you failed 1 test of 4.
#   Failed test '->die'
#   at t/validations/00_base_class.t line 51.
# Looks like you failed 1 test of 8.
```

lseppala commented:
@sungo ready for re-review

I addressed your comments in one commit and then re-worked loading the pre-configured validation plans in the other. Thanks for the nudge, it ended up a much better solution.
